Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp

```diff
@@ -0,0 +1,628 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WPEInputMethodContextWaylandV3.h"
+
+#include "WPEDisplayWaylandPrivate.h"
+#include "WPEViewWayland.h"
+
+#include "text-input-unstable-v3-client-protocol.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <wayland-client-protocol.h>
+#include <wtf/glib/WTFGType.h>
+#include <xkbcommon/xkbcommon.h>
+
+typedef struct _TextInputV3Global TextInputV3Global;
+
+struct _TextInputV3Global {
+    struct zwp_text_input_v3* textInput;
+
+    WPEInputMethodContext* current;
+
+    bool focused;
+
+    uint32_t serial;
+};
+
+typedef struct {
+    char* text;
+    int32_t cursorBegin;
+    int32_t cursorEnd;
+} Preedit;
+
+struct _WPEIMContextWaylandV3Private {
+    WPEView* view;
+
+    Preedit pendingPreedit;
+    Preedit currentPreedit;
+
+    char* pendingCommit;
+
+    struct {
+        int32_t x;
+        int32_t y;
+        int32_t width;
+        int32_t height;
+    } cursor_rect;
+
+    struct {
+        char* text;
+        int32_t cursorIndex;
+        int32_t anchorIndex;
+    } surrounding;
+
+    enum zwp_text_input_v3_change_cause textChangeCause;
+
+    struct {
+        uint32_t beforeLength;
+        uint32_t afterLength;
+    } pending_surrounding_delete;
+};
+
+WEBKIT_DEFINE_FINAL_TYPE(WPEIMContextWaylandV3, wpe_im_context_wayland_v3, WPE_TYPE_INPUT_METHOD_CONTEXT, WPEInputMethodContext)
```

Comment:
I think there's some common code betwen v1 and v3, so maybe we can design this differently to avoid code duplication.

Reply:
There are similarities. My thought was that it's better to keep them separated as they are unstable and a bit vague protocols